src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,510 @@
+// plus ValidTime seconds. This function simply checks that the current time
+// ('now') is before the end of the validity interval.
+func (dc *DelegatedCredential) IsExpired(start, now time.Time) bool {
+	end := start.Add(dc.ValidTime)
+	return !now.Before(end)
+}
+
+// InvalidTTL returns true if the credential's validity period is longer than the
+// maximum permitted. This is defined by the certificate's notBefore field
+// ('start') plus the ValidTime, minus the current time ('now').
+func (dc *DelegatedCredential) InvalidTTL(start, now time.Time) bool {
+	return dc.ValidTime > (now.Sub(start) + dcMaxTTL).Round(time.Second)
+}
+
+// credential stores the public components of a delegated crendetial.
+type credential struct {
+	validTime time.Duration
+	publicKey crypto.PublicKey
+	scheme    SignatureScheme
+}
+
+// DelegatedCredential stores the credential and its signature.
+type DelegatedCredential struct {
+	// The serialized form of the credential.
+	Raw []byte
+
+	// The amount of time for which the credential is valid. Specifically, the
+	// the credential expires 'ValidTime' seconds after the 'notBefore' of the
+	// delegation certificate. The delegator shall not issue delegated
+	// credentials that are valid for more than 7 days from their current time.
+	//
+	// When this data structure is serialized, this value is converted to a
+	// uint32 representing the duration in seconds.
+	ValidTime time.Duration
+
+	// The credential's public key.
+	PublicKey crypto.PublicKey
+
+	// The signature scheme associated with the credential public key.
+	publicKeyScheme SignatureScheme
+
+	// The signature scheme used to sign the credential.
+	Scheme SignatureScheme
+
+	// The credential's delegation.
+	Signature []byte
+}
+
+// marshalSubjectPublicKeyInfo returns a DER encoded SubjectPublicKeyInfo structure
+// (as defined in the X.509 standard) for the credential.
+// TODO: maybe we can move this as well to x509
+// TODO: add the other signatures flavors, as defined in common.go
+func (cred *credential) marshalPublicKeyInfo() ([]byte, error) {
+	switch cred.scheme {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512:
+		serializedPublicKey, err := x509.MarshalPKIXPublicKey(cred.publicKey)
+		if err != nil {
+			return nil, err
+		}
+		return serializedPublicKey, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported signature scheme: 0x%04x", cred.scheme)
+	}
+}
+
+// unmarshalPublicKeyInfo parses a DER encoded PublicKeyInfo
+// structure into a public key and its corresponding algorithm.
+// TODO: add the other signatures flavors, as defined in common.go
+func unmarshalPublicKeyInfo(serialized []byte) (crypto.PublicKey, SignatureScheme, error) {
+	publicKey, err := x509.ParsePKIXPublicKey(serialized)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	switch pk := publicKey.(type) {
+	case *ecdsa.PublicKey:
+		curveName := pk.Curve.Params().Name
+		if curveName == "P-256" {
```

Comment:
In [cfssl](https://github.com/cloudflare/cfssl/blob/2916a1fa6bf37b6d7a672a9be7e734871e08039c/transport/kp/key_provider.go#L193) algorithm names are forced to be lower case.  If we want to use the same algorithm names everywhere, then I think it's best to have our API be:

 - case insensitive in the sense that it would accept `Ed25519`, `ed25519`, `ED25519`, etc. as the same name.
 - case sensitive in the sense that `Scheme.Name()` always returns the right capatalization (`Ed25519`).

Reply:
ok let's follow that idea.